daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala

```diff
@@ -183,6 +202,35 @@ abstract class SequenceParserBase(
               //
               pstate.mpstate.moveOverOneGroupIndexOnly()
             }
+
+            val newLastChildNode = pstate.infoset.contents.lastOption
+            if (newLastChildNode != savedLastChildNode) {
+              // We have added at least one child to to this complex during
+              // this array loop.
+              //
+              // If the new child is a DIArray, we know this DIArray has at
+              // least one element, but we don't know if we actually added a
+              // new one in this loop or not. So just get the last array
+              // element and set it as final anyways. Note that we need a null
+              // check because if we didn't add an array element this go
+              // around, the last element may have already been walked and
+              // freed and so could now be null.
+              //
+              // If it's not a DIArray, that means it's just an optional
+              // simple/complex and that will get set final below where all
+              // other non-array elements get set as final.
+              newLastChildNode.get match {
+                case a: DIArray => {
+                  val lastArrayElem = a.contents.last
```

Comment:
Not about this line of code, but I realized here that we're pruning elements from arrays, but haven't changed the array representation to not be O(n) in the size of the array. Hence, for a very big array we'll still run into the limitation of max JVM array size, and it will always need to be in memory. If I am correct here then we should park the further enhancement to change the array representation. Whether that will prove important or not, I'm not sure. I imagine that pruning the array elements is the biggest win.

Reply:
Yeah, for super large arrays (not in physical size, but in count) we aren't freeing up the array memory until the entire array is finished. I decide to punt on that for now just because it adds extra complication about keeping track of the count and dealing with offsets. I'll add a bug to potentially add this as a future optimization.